README
Socket6

$Id: README,v 1.10 2005/08/27 16:33:09 ume Exp $

1. Abstract

This module supports getaddrinfo() and getnameinfo() to intend to
enable protocol independent programing.
If your environment supports IPv6, IPv6 related defines such as
AF_INET6 are included.

2. Install

Extract archive file.
Change current directory to Socket6-x.xx ("x.xx" is version number).
Type "perl Makefile.PL" to create Makefile.
Type "make" to build.
Type "make install" to install.

3. How to use

See "perldoc Socket6".

4. Acknowledgments

This package is mainly depend on perl5.005_55-v6-19990721 and
ssh-1.2.27-IPv6-1.5.  perl5.005_55-v6-19990721 was written by KAME
Project and ssh-1.2.27-IPv6-1.5 was written by KIKUCHI Takahiro
<kick@kyoto.wide.ad.jp>.  Thank you very much.

5. Copyright

Copyright (C) 2000-2005 Hajimu UMEMOTO <ume@mahoroba.org>.
All rights reserved.

Socket6.pm and Socket6.xs are based on perl5.005_55-v6-19990721
written by KAME Project.

gai.h, getaddrinfo.c and getnameinfo.c are based on
ssh-1.2.27-IPv6-1.5 written by KIKUCHI Takahiro
<kick@kyoto.wide.ad.jp>.

Copyright (C) 1995, 1996, 1997, 1998, and 1999 WIDE Project.
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions
are met:
1. Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
3. Neither the name of the project nor the names of its contributors
   may be used to endorse or promote products derived from this software
   without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
SUCH DAMAGE.


ChangeLog
2013-12-08  Hajimu UMEMOTO  <ume@mahoroba.org>

	* Socket6.pm: Bump version number to 0.24.

	* Makefile.PL: Fixed [cpan #77665] pass perl cc and ccflags to
	configure

	Submitted by:	rurban@cpan.org

	* Makefile.PL (configure_mswin32): Fixed [cpan #65710]
	undefined ref to inet_ntop,inet_pton on Windows XP.

	Submitted by:	APOCAL@cpan.org

	* t/use.t: Fixed [cpan #75211] test redefinition warnings

	Submitted by:	Manoj Kumar <mkmetw@hotmail.com>

2008-11-02  Hajimu UMEMOTO  <ume@mahoroba.org>

	* Socket6.pm: Bump version number to 0.23.

	* Makefile.PL (configure_mswin32): Added a conditional #define for
	HAVE_INET_(PTON|NTOP) to make the module.
	Change was necessary because the SDK for Vista now includes
	inet_(pton|ntop) in ws2tcpip.h

	Submitted by:	"David M. Town" <dtown@cpan.org>

	* Makefile.PL (configure_mswin32): Added a line to #define
	HAVE_SOCKLEN_T since it is defined for MSWin32

	Submitted by:	"David M. Town" <dtown@cpan.org>

	* Socket6.pm: Added @EXPORT_OK to export AF_INET6 and PF_INET6 if
	requested.  This allows the use of the Socke6 module without
	having to perform the same conditional check that was added to
	avoid conflicts with the Socket module default exports (i.e. use
	Socket6 qw(AF_INET6 inet_pton inet_ntop); )

	Submitted by:	"David M. Town" <dtown@cpan.org>

	* Socket6.pm (AUTOLOAD): Explicitly set $! = 0 before calling
	constant().  This should not be required since errno was being set
	to 0 in the C function, but Perl 5.10.0 Win32 Vista needed this to
	be done or all constants reported unsupported.

	Submitted by:	"David M. Town" <dtown@cpan.org>

2008-08-18  Hajimu UMEMOTO  <ume@mahoroba.org>

	* Socket6.pm: Use strict environment.

	* Socket6.pm: Bump version number to 0.22.

	* Socket6.pm: Make AF_INET6 check actually work.

	Submitted by:	Tatsuya Noda <topia@clovery.jp>

	* Makefile.PL: Make sure to include config.pl in current
	directory.

2008-08-17  Hajimu UMEMOTO  <ume@mahoroba.org>

	* Socket6.pm: Bump version number to 0.21.

	* Socket6.xs: Make it buildable on the following environment by
	defining WINVER as 0x0501:
		- Windows XP SP3
		- ActivePerl-5.10.0.1003-MSWin32-x86-285500
		- MinGW-5.1.4
		- nmake 9.00.21022.08 (shipped with VisualStudio 2008 Express)

	Submitted by:	"IWAMURO Motonori" <vmi@nifty.com>

2008-02-16  Hajimu UMEMOTO  <ume@mahoroba.org>

	* Socket6.pm: Bump version number to 0.20.

	* Socket6.pm: Make the AF_INET6 and PF_INET6 exports optional.
	They are exported from Socket on 5.10.0, and it contlicts with
	Socket6.

	PR:		http://rt.cpan.org/Public/Bug/Display.html?id=32362
	Reported by:	"Shlomi Fish via RT" <bug-Socket6@rt.cpan.org>
	Submitted by:	"KONDOU, Kazuhiro" <kazuhiro@alib.jp>

2008-01-27  Hajimu UMEMOTO  <ume@mahoroba.org>

	* Socket6.xs, Socket6.pm: Make getaddrinfo() return the string
	error or the numeric value according to the context.
	Make getnameinfo() return error value.
	Export the EAI_* error constants.

	Submitted by:	Paul LeoNerd Evans <leonerd@leonerd.org.uk>

	* Socket6.xs: FreeBSD 7 and later don't have KAME IPsec anymore.

	* Socket6.xs: gai_strerror() returns `const char *'.

2005-08-28  Hajimu UMEMOTO  <ume@mahoroba.org>

	* Socket6.pm: Bump version number to 0.19.

	* Socket6.xs: the "sin6_scope_id" field of the "sockaddr_in6"
	structure is expected to passed and is returned in host byte
	order.

	Submitted by:	"Town, David" <David.Town@marconi.com>

2005-03-28  Hajimu UMEMOTO  <ume@mahoroba.org>

	* Socket6.pm: Bump version number to 0.18.

	* getaddrinfo.c (getaddrinfo): Support for AI_NUMERICSERV.

	* aclocal.m4: use Autoconf 2.59.

2004-07-14  Hajimu UMEMOTO  <ume@mahoroba.org>

	* Socket6.pm: Fix documentation of the gethostbyname2.  The
	gethostbyname2 function takes a FAMILY, not a SERVNAME.

		Reported by:	Bernard Massot <bernard@massot.ath.cx>

2004-03-24  Hajimu UMEMOTO  <ume@mahoroba.org>

	* Socket6.pm: Bump version number to 0.17.

	* Socket6.xs: Shut up signed/unsigned mismatch warning from
	Microsoft C++.

		Reported by:	"David Town" <dtown@cpan.org>

	* Makefile.PL: CC is not a valid parameter for WriteMakefile().
	Use CONFIGURE instead.

		Submitted by:	"David Town" <dtown@cpan.org>

	* Socket6.pm: Correct typos (inaddr6_* -> in6addr_*).

		Submitted by:	"David Town" <dtown@cpan.org>

2004-03-23  Hajimu UMEMOTO  <ume@mahoroba.org>

	* Provide substitutes for systems the lack support for inet_ntop()
	and inet_pton().

		Submitted by:	"David Town" <dtown@cpan.org>

	* Add support for MSWin32.  It was tested with the following
	configuration:

	  - ActivePerl 5.8.0 (805)
	  - Windows 2000
	  - Microsoft IPv6 Technology Preview for Windows 2000 (not needed
	    for XP)
	    (http://msdn.microsoft.com/downloads/sdks/platform/tpipv6.asp)
	  - Microsoft Visual C++ v6.0
	  - Microsoft Platform SDK February 2003 (must have at least
	    August 2001)
	    (http://www.microsoft.com/msdownload/platformsdk/sdkupdate/downlevel.htm)

		Submitted by:	"David Town" <dtown@cpan.org>

	* t/use.t: Pass an empty string as the SERVICENAME to
	getaddrinfo().  MSWin32's getaddrinfo() doesn't like "0" as a
	numeric servname.

		Submitted by:	"David Town" <dtown@cpan.org>

2004-02-21  Hajimu UMEMOTO  <ume@mahoroba.org>

	* Socket6.pm: Bump version number to 0.16.

	* aclocal.m4: inet_ntop and inet_pton are in libnsl on Solaris.

	Submitted by:	Mike McCauley <mikem@open.com.au>

2004-01-05  Hajimu UMEMOTO  <ume@mahoroba.org>

	* Socket6.pm: Bump version number to 0.14.

	* gailookup.pl.in: Though some systems has NI_WITHSCOPEID, RFC
	2553 doesn't mention it, and RFC 3493 deprecated it.  So, make use
	of NI_WITHSCOPEID optional.

	* gailookup.pl.in: Add -S option which suppress scopeid.  RFC 3493
	system always adds scopeid in a result of getnameinfo().

	* Socket6.xs: Make pack_sockaddr_in6() and pack_sockaddr_in6_all()
	actual work on systems which have sin6_len in struct sockaddr_in6.

2004-01-04  Hajimu UMEMOTO  <ume@mahoroba.org>

	* aclocal.m4: getaddrinfo() is a macro on some systems.


Makefile.PL
# Copyright (C) 2000-2013 Hajimu UMEMOTO <ume@mahoroba.org>.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 3. Neither the name of the project nor the names of its contributors
#    may be used to endorse or promote products derived from this software
#    without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.

# $Id: Makefile.PL,v 1.12 2013/12/08 17:21:56 ume Exp $

use Config;
use ExtUtils::MakeMaker;

if ($^O eq 'MSWin32') {
    configure_mswin32();
} else {
    my $path_perl = ($^X =~ m!^/!o) ? $^X : `which $^X`;
    my $cmd = "CC='$Config{cc}' CFLAGS='$Config{ccflags}' ./configure";
    $cmd .= " --with-perl=$path_perl" if ($path_perl);
    system($cmd);
}

require './config.pl';

WriteMakefile(
    NAME	 => 'Socket6',
    VERSION_FROM => 'Socket6.pm',
    XSPROTOARG	 => '-noprototypes',	# XXX remove later?
    PM		 => {'Socket6.pm' => '$(INST_LIBDIR)/Socket6.pm'},
    CONFIGURE	 => sub { { CC => "${cc}" } },
    LIBS	 => ["${libs}"],
    realclean	 => {FILES => 'config.cache config.h config.log config.pl config.status gailookup.pl'},
);

sub configure_mswin32
{
    open(IN, '<config.pl.in') || die "Failed to open file 'config.pl.in' [$!]";
    open(OUT, '>config.pl')   || die "Failed to open file 'config.pl' [$!]";

    while (<IN>) {
	if (/\$cc/) {
	    printf(OUT "\$cc = '%s';\n", $Config::Config{cc});
	} elsif (/\$libs/) {
	    printf(OUT "\$libs = 'ws2_32.lib';\n");
	} else {
	    print OUT;
	}
    }

    close(OUT) || die "Failed to close file 'config.pl' [$!]";
    close(IN)  || die "Failed to close file 'config.pl.in' [$!]";

    open(IN, '<config.h.in') || die "Failed to open file 'config.h.in' [$!]";
    open(OUT, '>config.h')   || die "Failed to open file 'config.h' [$!]";

    print(OUT "#ifndef NTDDI_LONGHORN\n");
    print(OUT "#  define NTDDI_LONGHORN 0x06000000\n");
    print(OUT "#endif\n");

    while (<IN>) {
	if (/HAVE_PL_SV_UNDEF/) {
	    print(OUT "#define HAVE_PL_SV_UNDEF 1\n");
	} elsif (/HAVE_GETADDRINFO/) {
	    print(OUT "#define HAVE_GETADDRINFO 1\n");
	} elsif (/HAVE_GETNAMEINFO/) {
	    print(OUT "#define HAVE_GETNAMEINFO 1\n");
	} elsif (/HAVE_SOCKADDR_IN6_SIN6_SCOPE_ID/) {
	    print(OUT "#define HAVE_SOCKADDR_IN6_SIN6_SCOPE_ID 1\n");
        } elsif (/HAVE_INET_(PTON|NTOP)/) {
	    print(OUT "#if (NTDDI_VERSION >= NTDDI_LONGHORN)\n");
	    print(OUT "  #define HAVE_INET_$1 1\n");
	    print(OUT "#else\n");
	    print(OUT "  #undef HAVE_INET_$1\n");
	    print(OUT "#endif\n");
	} elsif (/HAVE_SOCKLEN_T/) {
	    print(OUT "#define HAVE_SOCKLEN_T 1\n");
	} else {
	    print OUT;
	}
    }

    close(OUT) || die "Failed to close file 'config.h' [$!]";
    close(IN)  || die "Failed to close file 'config.h.in' [$!]";
}


Socket6.pm
# Copyright (C) 2000-2008 Hajimu UMEMOTO <ume@mahoroba.org>.
# All rights reserved.
#
# This module is besed on perl5.005_55-v6-19990721 written by KAME
# Project.
#
# Copyright (C) 1995, 1996, 1997, 1998, and 1999 WIDE Project.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
# 3. Neither the name of the project nor the names of its contributors
#    may be used to endorse or promote products derived from this software
#    without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.

# $Id: Socket6.pm,v 1.46 2013/12/07 18:49:55 ume Exp $

package Socket6;

use strict;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS $AUTOLOAD);
$VERSION = "0.24";

=head1 NAME

Socket6 - IPv6 related part of the C socket.h defines and structure manipulators

=head1 SYNOPSIS

    use Socket;
    use Socket6;

    @res = getaddrinfo('hishost.com', 'daytime', AF_UNSPEC, SOCK_STREAM);
    $family = -1;
    while (scalar(@res) >= 5) {
	($family, $socktype, $proto, $saddr, $canonname, @res) = @res;

	($host, $port) = getnameinfo($saddr, NI_NUMERICHOST | NI_NUMERICSERV);
	print STDERR "Trying to connect to $host port port $port...\n";

	socket(Socket_Handle, $family, $socktype, $proto) || next;
        connect(Socket_Handle, $saddr) && last;

	close(Socket_Handle);
	$family = -1;
    }

    if ($family != -1) {
	print STDERR "connected to $host port port $port\n";
    } else {
	die "connect attempt failed\n";
    }

=head1 DESCRIPTION

This module provides glue routines to the various IPv6 functions.

If you use the Socket6 module,
be sure to specify "use Socket" as well as "use Socket6".

Functions supplied are:

=item inet_pton FAMILY, TEXT_ADDRESS

    This function takes an IP address in presentation (or string) format
    and converts it into numeric (or binary) format.
    The type of IP address conversion (IPv4 versus IPv6) is controlled
    by the FAMILY argument.

=item inet_ntop FAMILY, BINARY_ADDRESS

    This function takes an IP address in numeric (or binary) format
    and converts it into presentation (or string) format
    The type of IP address conversion (IPv4 versus IPv6) is controlled
    by the FAMILY argument.

=item pack_sockaddr_in6 PORT, ADDR

    This function takes two arguments: a port number, and a 16-octet
    IPv6 address structure (as returned by inet_pton()).
    It returns the sockaddr_in6 structure with these arguments packed
    into their correct fields, as well as the AF_INET6 family.
    The other fields are not set and their values should not be relied upon.

=item pack_sockaddr_in6_all PORT, FLOWINFO, ADDR, SCOPEID

    This function takes four arguments: a port number, a 16-octet
    IPv6 address structure (as returned by inet_pton), any
    special flow information, and any specific scope information.
    It returns a complete sockaddr_in6 structure with these arguments packed
    into their correct fields, as well as the AF_INET6 family.

=item unpack_sockaddr_in6 NAME

    This function takes a sockaddr_in6 structure (as returned by
    pack_sockaddr_in6()) and returns a list of of two elements:
    the port number and the 16-octet IP address.
    This function will croak if it determines it has not been
    passed an IPv6 structure.

=item unpack_sockaddr_in6_all NAME

    This function takes a sockaddr_in6 structure (as returned by
    pack_sockaddr_in6()) and returns a list of of four elements:
    the port number, the flow information, the 16-octet IP address,
    and the scope information.
    This function will croak if it determines it has not been
    passed an IPv6 structure.

=item gethostbyname2 HOSTNAME, FAMILY

=item getaddrinfo NODENAME, SERVICENAME, [FAMILY, SOCKTYPE, PROTOCOL, FLAGS]

    This function converts node names to addresses and service names
    to port numbers.
    If the NODENAME argument is not a false value,
    then a nodename to address lookup is performed;
    otherwise a service name to port number lookup is performed.
    At least one of NODENAME and SERVICENAME must have a true value.

    If the lookup is successful, a list consisting of multiples of
    five elements is returned.
    Each group of five elements consists of the address family,
    socket type, protocol, 16-octet IP address, and the canonical
    name (undef if the node name passed is already the canonical name).

    The arguments FAMILY, SOCKTYPE, PROTOCOL, and FLAGS are all optional.

    This function will croak if it determines it has not been
    passed an IPv6 structure.

    If the lookup is unsuccessful, the function returns a single scalar.
    This will contain the string version of that error in string context,
    and the numeric value in numeric context.

=item getnameinfo NAME, [FLAGS]

    This function takes a socket address structure. If successful, it returns
    two strings containing the node name and service name.

    The optional FLAGS argument controls what kind of lookup is performed.

    If the lookup is unsuccessful, the function returns a single scalar.
    This will contain the string version of that error in string context,
    and the numeric value in numeric context.

=item getipnodebyname HOST, [FAMILY, FLAGS]

    This function takes either a node name or an IP address string
    and performs a lookup on that name (or conversion of the string).
    It returns a list of five elements: the canonical host name,
    the address family, the length in octets of the IP addresses
    returned, a reference to a list of IP address structures, and
    a reference to a list of aliases for the host name.

    The arguments FAMILY and FLAGS are optional.
    Note: This function does not handle IPv6 scope identifiers,
    and should be used with care.
    And, this function was deprecated in RFC3493.
    The getnameinfo function should be used instead.

=item getipnodebyaddr FAMILY, ADDRESS

    This function takes an IP address family and an IP address structure
    and performs a reverse lookup on that address.
    It returns a list of five elements: the canonical host name,
    the address family, the length in octets of the IP addresses
    returned, a reference to a list of IP address structures, and
    a reference to a list of aliases for the host name.

    Note: This function does not handle IPv6 scope identifiers,
    and should be used with care.
    And, this function was deprecated in RFC3493.
    The getaddrinfo function should be used instead.

=item gai_strerror ERROR_NUMBER

    This function returns a string corresponding to the error number
    passed in as an argument.

=item in6addr_any

    This function returns the 16-octet wildcard address.

=item in6add_loopback

    This function returns the 16-octet loopback address.

=over

=back

=cut

use Carp;

use base qw(Exporter DynaLoader);

@EXPORT = qw(
	inet_pton inet_ntop pack_sockaddr_in6 pack_sockaddr_in6_all
	unpack_sockaddr_in6 unpack_sockaddr_in6_all sockaddr_in6
	gethostbyname2 getaddrinfo getnameinfo
	in6addr_any in6addr_loopback
	gai_strerror getipnodebyname getipnodebyaddr
	AI_ADDRCONFIG
	AI_ALL
	AI_CANONNAME
	AI_NUMERICHOST
	AI_NUMERICSERV
	AI_DEFAULT
	AI_MASK
	AI_PASSIVE
	AI_V4MAPPED
	AI_V4MAPPED_CFG
	EAI_ADDRFAMILY
	EAI_AGAIN
	EAI_BADFLAGS
	EAI_FAIL
	EAI_FAMILY
	EAI_MEMORY
	EAI_NODATA
	EAI_NONAME
	EAI_SERVICE
	EAI_SOCKTYPE
	EAI_SYSTEM
	EAI_BADHINTS
	EAI_PROTOCOL
	IP_AUTH_TRANS_LEVEL
	IP_AUTH_NETWORK_LEVEL
	IP_ESP_TRANS_LEVEL
	IP_ESP_NETWORK_LEVEL
	IPPROTO_IP
	IPPROTO_IPV6
	IPSEC_LEVEL_AVAIL
	IPSEC_LEVEL_BYPASS
	IPSEC_LEVEL_DEFAULT
	IPSEC_LEVEL_NONE
	IPSEC_LEVEL_REQUIRE
	IPSEC_LEVEL_UNIQUE
	IPSEC_LEVEL_USE
	IPV6_AUTH_TRANS_LEVEL
	IPV6_AUTH_NETWORK_LEVEL
	IPV6_ESP_NETWORK_LEVEL
	IPV6_ESP_TRANS_LEVEL
	NI_NOFQDN
	NI_NUMERICHOST
	NI_NAMEREQD
	NI_NUMERICSERV
	NI_DGRAM
	NI_WITHSCOPEID
);
push @EXPORT, qw(AF_INET6) unless defined eval {Socket::AF_INET6()};
push @EXPORT, qw(PF_INET6) unless defined eval {Socket::PF_INET6()};

@EXPORT_OK = qw(AF_INET6 PF_INET6);

%EXPORT_TAGS = (
    all     => [@EXPORT],
);

sub sockaddr_in6 {
    if (wantarray) {
	croak "usage:   (port,iaddr) = sockaddr_in6(sin_sv)" unless @_ == 1;
        unpack_sockaddr_in6(@_);
    } else {
	croak "usage:   sin_sv = sockaddr_in6(port,iaddr))" unless @_ == 2;
        pack_sockaddr_in6(@_);
    }
}

sub AUTOLOAD {
    my($constname);
    ($constname = $AUTOLOAD) =~ s/.*:://o;
    $! = 0;
    my $val = constant($constname, @_ ? $_[0] : 0);
    if ($! != 0) {
	croak "Your vendor has not defined Socket macro $constname, used";
    }
    eval "sub $AUTOLOAD { $val }";
    goto &$AUTOLOAD;
}

bootstrap Socket6 $VERSION;

1;
